### ISSUE TITLE
Promise<T|undefined>.then(...) accepts wrong type of argument

### ISSUE THREAD

**Issue by james-boxem** [microsoft/TypeScript#63134]
### 🔎 Search Terms

typecheck promise parameter coerce undefined type error

### 🕗 Version & Regression Information

I have tested this in v5.9.3 (latest version at time of writing)

### 💻 Code

minimal reproduction of issue:
```ts
// this should not pass type checks! param `x` may receive `undefined`!
new Promise<number | undefined>(resolve => resolve(undefined)).then((x: number) => console.log(x));


// if we define `f` this way,
const f = (x: number) => console.log(x);

// this is detected as an error
new Promise<number | undefined>(resolve => resolve(undefined)).then(x => f(x));

// but this is not! 😬
new Promise<number | undefined>(resolve => resolve(undefined)).then(f);
                  



```

### 🙁 Actual behavior

I think it's fairly self explanatory since the example is extremely minimal.

### 🙂 Expected behavior

I think it's fairly self explanatory since the example is extremely minimal.

**Comment by MartinJohns** [comment]
You didn't fill out the issue template completely.



> // this should not pass type checks! param `x` may receive `undefined`!
> new Promise<number | undefined>(resolve => resolve(undefined)).then((x: number) => console.log(x));

Well, what can I tell you, it **doesn't** pass type checks! It produces an error.

> // but this is not! 😬
> new Promise<number | undefined>(resolve => resolve(undefined)).then(f);

But.. it does! It does produce an error!

**Comment by james-boxem** [comment]
hm, it is definitely passing type checks for me.
I found an even more general form of this issue, also passing type checks:

```ts
// this should not pass type checks, but it does
const f: (x: undefined | number) => void = (x: number) => {};
```

**Comment by james-boxem** [comment]
OK, I now see the error if I turn on "strict" mode. But when "strict" is disabled but "strictNullChecks" = `true`, this passes type checks

**Comment by james-boxem** [comment]
OK, I can turn on "strictFunctionTypes" and the error is caught. Not a bug, my bad.

**Comment by MartinJohns** [comment]
> But when "strict" is disabled but "strictNullChecks" = `true`, this passes type checks

Which is correct. With this option turned off there is no type `undefined`, so having `number | undefined` is the same as `number`.
